Document llvm passes in more depth (#50319)
JuliaLang/julia: doc/src/devdocs/llvm-passes.md

Reviewer: Don't you need to add this page to `doc/make.jl` to actually add it to the docs?

### FinalGCLowering

* Filename: `llvm-final-gc-lowering.cpp`
* Class Name: `FinalLowerGCPass`
* Opt Name: `module(FinalLowerGC)`

This pass lowers a few last intrinsics to their final form targeting functions in the `libjulia` library. Separating this from `LateGCLowering` enables other backends (GPU compilation) to supply their own custom lowerings for these intrinsics, enabling the Julia pipeline to be used on those backends as well.

### LowerHandlers

* Filename: `llvm-lower-handlers.cpp`
* Class Name: `LowerExcHandlersPass`
* Opt Name: `function(LowerExcHandlers)`

This pass lowers exception handling intrinsics into calls to runtime functions that are actually called when handling exceptions.

### RemoveNI

* Filename: `llvm-remove-ni.cpp`
* Class Name: `RemoveNIPass`
* Opt Name: `module(RemoveNI)`

> This pass removes the non-integral address spaces from the module's datalayout string. This enables the backend to lower Julia's custom address spaces directly to machine code, without a costly rewrite of every pointer operation to address space 0.

Reviewer 2: I assume you meant this in the reverse? "This enables buggy backends to still lower Julia's custom address spaces to machine code, albeit with a costly rewrite of every pointer operation to address space 0."

Reviewer 2: Oh, sorry, I was mixing this up this pass with RemoveAddrspaces

### SIMDLoop

* Filename: `llvm-simdloop.cpp`
* Class Name: `LowerSIMDLoopPass`
* Opt Name: `module(LowerSIMDLoop)`

> This pass acts as the main driver of the `@simd` annotation. Codegen inserts a call to a marker intrinsic (`julia.simdloop`), which this pass uses to identify loops that were originally marked with `@simd`. Then, this pass looks for chains of floating point operations to mark as fast floating point operations. If the loop was annotated with `ivdep` as well, then the pass marks the loop as having no loop-carried dependencies.

Reviewer 2: ```suggestion
This pass acts as the main driver of the `@simd` annotation. Codegen inserts a call to a marker intrinsic (`julia.simdloop`), which this pass uses to identify loops that were originally marked with `@simd`. Then, this pass looks for a chain of floating point operations that form a reduce and marks adds fast math to allow reassociation (and thus vectorization). This pass does not preserve either loop information nor inference correctness, so it may violate Julia semantics in surprising ways. If the loop was annotated with `ivdep` as well, then the pass marks the loop as having no loop-carried dependencies (the resulting behavior is undefined if the user annotation was incorrect or gets applied to the wrong loop).
```

Reviewer 3: Also since #49405 we only add a limited set of fast math flags.

### LowerPTLS

* Filename: `llvm-ptls.cpp`
* Class Name: `LowerPTLSPass`
* Opt Name: `module(LowerPTLSPass)`

> This pass lowers thread-local Julia intrinsics to assembly instructions. Julia relies on thread-local storage for garbage collection and multithreading task scheduling. When compiling code for system images and package images, this pass replaces calls to intrinsics with loads from global variables that are initialized at load time.

Reviewer 2: "with loads from TLS, or the swiftself argument"?

> This pass lowers thread-local Julia intrinsics to assembly instructions. Julia relies on thread-local storage for garbage collection and multithreading task scheduling. When compiling code for system images and package images, this pass replaces calls to intrinsics with loads from global variables that are initialized at load time.

Author: ```suggestion
This pass lowers thread-local Julia intrinsics to assembly instructions. Julia relies on thread-local storage for garbage collection and multithreading task scheduling. When compiling code for system images and package images, this pass replaces calls to intrinsics with loads from global variables that are initialized at load time.

If codegen produces a function with a `swiftself` argument and calling convention, this pass assumes the `swiftself` argument is the pgcstack and will replace the intrinsics with that argument. Doing so provides speedups on architectures that have slow thread local storage accesses.
```

### RemoveAddrspaces

* Filename: `llvm-remove-addrspaces.cpp`
* Class Name: `RemoveAddrspacesPass`
* Opt Name: `module(RemoveAddrspaces)`

This pass renames pointers in one address space to another address space. This is used to remove Julia-specific address spaces from LLVM IR.

### RemoveJuliaAddrspaces

* Filename: `llvm-remove-addrspaces.cpp`
* Class Name: `RemoveJuliaAddrspacesPass`
* Opt Name: `module(RemoveJuliaAddrspaces)`

This pass removes Julia-specific address spaces from LLVM IR. It is mostly used for displaying LLVM IR in a less cluttered format. Internally, it is implemented off the RemoveAddrspaces pass.

### Multiversioning

* Filename: `llvm-multiversioning.cpp`
* Class Name: `MultiVersioningPass`
* Opt Name: `module(JuliaMultiVersioning)`

> This pass performs modifications to a module to create functions that are optimized for running on different architectures (see sysimg.md and pkgimg.md for more details). Implementation-wise, it clones functions and applies different target-specific attributes to them to allow the optimizer to use advanced features such as vectorization and instruction scheduling for that platform. It also creates some infrastructure to enable the Julia image loader to select the appropriate version of the function to call based on the architecture the loader is running on. The target-specific attributes are controlled by the `JULIA_CPU_TARGET` environment variable.

Reviewer 2: It is controlled by the Module flags `julia.mv.enable` and `julia.mv.specs`, though those values may be influenced by JULIA_CPU_TARGET.

> This pass performs modifications to a module to create functions that are optimized for running on different architectures (see sysimg.md and pkgimg.md for more details). Implementation-wise, it clones functions and applies different target-specific attributes to them to allow the optimizer to use advanced features such as vectorization and instruction scheduling for that platform. It also creates some infrastructure to enable the Julia image loader to select the appropriate version of the function to call based on the architecture the loader is running on. The target-specific attributes are controlled by the `JULIA_CPU_TARGET` environment variable.

Author: ```suggestion
This pass performs modifications to a module to create functions that are optimized for running on different architectures (see sysimg.md and pkgimg.md for more details). Implementation-wise, it clones functions and applies different target-specific attributes to them to allow the optimizer to use advanced features such as vectorization and instruction scheduling for that platform. It also creates some infrastructure to enable the Julia image loader to select the appropriate version of the function to call based on the architecture the loader is running on. The target-specific attributes are controlled by the `julia.mv.specs` module flag, which during compilation is derived from the `JULIA_CPU_TARGET` environment variable. The pass must also be enabled by providing a `julia.mv.enable` module flag with a value of 1.
```

### FinalGCLowering

* Filename: `llvm-final-gc-lowering.cpp`
* Class Name: `FinalLowerGCPass`
* Opt Name: `module(FinalLowerGC)`

This pass lowers a few last intrinsics to their final form targeting functions in the `libjulia` library. Separating this from `LateGCLowering` enables other backends (GPU compilation) to supply their own custom lowerings for these intrinsics, enabling the Julia pipeline to be used on those backends as well.

### LowerHandlers

* Filename: `llvm-lower-handlers.cpp`
* Class Name: `LowerExcHandlersPass`
* Opt Name: `function(LowerExcHandlers)`

This pass lowers exception handling intrinsics into calls to runtime functions that are actually called when handling exceptions.

### RemoveNI

* Filename: `llvm-remove-ni.cpp`
* Class Name: `RemoveNIPass`
* Opt Name: `module(RemoveNI)`

This pass removes the non-integral address spaces from the module's datalayout string. This enables the backend to lower Julia's custom address spaces directly to machine code, without a costly rewrite of every pointer operation to address space 0.

### SIMDLoop

* Filename: `llvm-simdloop.cpp`
* Class Name: `LowerSIMDLoopPass`
* Opt Name: `module(LowerSIMDLoop)`

> This pass acts as the main driver of the `@simd` annotation. Codegen inserts a call to a marker intrinsic (`julia.simdloop`), which this pass uses to identify loops that were originally marked with `@simd`. Then, this pass looks for a chain of floating point operations that form a reduce and marks adds fast math to allow reassociation (and thus vectorization). This pass does not preserve either loop information nor inference correctness, so it may violate Julia semantics in surprising ways. If the loop was annotated with `ivdep` as well, then the pass marks the loop as having no loop-carried dependencies (the resulting behavior is undefined if the user annotation was incorrect or gets applied to the wrong loop).

Author: ```suggestion
This pass acts as the main driver of the `@simd` annotation. Codegen inserts a call to a marker intrinsic (`julia.simdloop`), which this pass uses to identify loops that were originally marked with `@simd`. Then, this pass looks for a chain of floating point operations that form a reduce and adds the `contract` and `reassoc` fast math flags to allow reassociation (and thus vectorization). This pass does not preserve either loop information nor inference correctness, so it may violate Julia semantics in surprising ways. If the loop was annotated with `ivdep` as well, then the pass marks the loop as having no loop-carried dependencies (the resulting behavior is undefined if the user annotation was incorrect or gets applied to the wrong loop).
```
